File: crs-transformation-adapter-test/src/test/java/com/programmerare/crsTransformations/CrsTransformationAdapterLeafFactoryTest.java
package com.programmerare.crsTransformations;

import com.programmerare.crsConstants.constantsByAreaNameNumber.v9_5_4.EpsgNumber;
import com.programmerare.crsTransformationAdapterGeoPackageNGA.CrsTransformationAdapterGeoPackageNGA;
import com.programmerare.crsTransformationAdapterGeoTools.CrsTransformationAdapterGeoTools;
import com.programmerare.crsTransformationAdapterGooberCTL.CrsTransformationAdapterGooberCTL;
import com.programmerare.crsTransformationAdapterOrbisgisCTS.CrsTransformationAdapterOrbisgisCTS;
import com.programmerare.crsTransformationAdapterProj4J.CrsTransformationAdapterProj4J;
import com.programmerare.crsTransformations.coordinate.CrsCoordinate;
import com.programmerare.crsTransformations.coordinate.CrsCoordinateFactory;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

import static org.hamcrest.CoreMatchers.hasItem;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.greaterThanOrEqualTo;
import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class CrsTransformationAdapterLeafFactoryTest {

    public final static int EXPECTED_NUMBER_OF_ADAPTER_LEAF_IMPLEMENTATIONS = 5;

    private static List<String> actualClassNamesForAllKnownImplementations;
    
    @BeforeAll
    static void beforeAll() {
        actualClassNamesForAllKnownImplementations = Arrays.asList(
            CrsTransformationAdapterGooberCTL.class.getName(),
            CrsTransformationAdapterGeoPackageNGA.class.getName(),
            CrsTransformationAdapterGeoTools.class.getName(),
            CrsTransformationAdapterOrbisgisCTS.class.getName(),
            CrsTransformationAdapterProj4J.class.getName()
        );        
    }

    @Test
    void listOfNonClassNamesForAdapters_shouldNotBeRecognizedAsAdapters() {
        final List<String> stringsNotBeingClassNameForAnyAdapter = Arrays.asList(
            null,
            "",
            "  ",
            " x ",
            "abc",

            // this test class i.e. the below "this" does not imlpement the interface so therefore assertFalse below
            this.getClass().getName()
        );

        for (String stringNotBeingClassNameForAnyAdapter : stringsNotBeingClassNameForAnyAdapter) {
            assertFalse(
                CrsTransformationAdapterLeafFactory.isCrsTransformationAdapter(stringNotBeingClassNameForAnyAdapter),
                "Should not have been recognized as adapter : " +  stringNotBeingClassNameForAnyAdapter
            );
        }        
    }

    @Test
    void listOfHardcodedClassnames_shouldBeCrsTransformationAdapters() {
        final List<String> hardcodedClassNamesForAllKnownImplementations = CrsTransformationAdapterLeafFactory.getClassNamesForAllKnownImplementations();
        for (String hardcodedClassNameForKnownImplementation : hardcodedClassNamesForAllKnownImplementations) {
            assertTrue(
                CrsTransformationAdapterLeafFactory.isCrsTransformationAdapter(hardcodedClassNameForKnownImplementation),
                "Name of failing class: " +  hardcodedClassNameForKnownImplementation
            );
        }
    }
    @Test
    void listOfHardcodedClassnames_shouldBeCreateableAsNonNullCrsTransformationAdapters() {
        final List<String> hardcodedClassNamesForAllKnownImplementations = CrsTransformationAdapterLeafFactory.getClassNamesForAllKnownImplementations();
        for (String hardcodedClassNameForKnownImplementation : hardcodedClassNamesForAllKnownImplementations) {
            CrsTransformationAdapter crsTransformationAdapter = CrsTransformationAdapterLeafFactory.createCrsTransformationAdapter(hardcodedClassNameForKnownImplementation);
            verifyThatTheCreatedAdapterIsRealObject(crsTransformationAdapter);
        }
    }

    private void verifyThatTheCreatedAdapterIsRealObject(CrsTransformationAdapter crsTransformationAdapter) {
        assertNotNull(crsTransformationAdapter);
        // below trying to use the created object to really make sure it works
        CrsCoordinate coordinateWgs84 = CrsCoordinateFactory.createFromYNorthingLatitudeAndXEastingLongitude(59.330231, 18.059196, EpsgNumber.WORLD__WGS_84__4326);
        CrsTransformationResult resultSweref99 = crsTransformationAdapter.transform(coordinateWgs84, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        assertNotNull(resultSweref99);
        assertTrue(resultSweref99.isSuccess());
    }

    @Test
    void listOfKnownInstances_shouldOnlyContainNonNullObjectsAndTheNumberOfItemsShouldBeAtLeastFive() {
        List<CrsTransformationAdapter> list = CrsTransformationAdapterLeafFactory.getInstancesOfAllKnownAvailableImplementations();
        assertThat(list.size(), greaterThanOrEqualTo(EXPECTED_NUMBER_OF_ADAPTER_LEAF_IMPLEMENTATIONS));
        for (CrsTransformationAdapter crsTransformationAdapter : list) {
            assertNotNull(crsTransformationAdapter);
        }
    }

    @Test
    void listOfHardcodedClassnames_shouldCorrespondToActualClassNames() {
       final List<String> hardcodedClassNamesForAllKnownImplementations = CrsTransformationAdapterLeafFactory.getClassNamesForAllKnownImplementations();
        assertEquals(EXPECTED_NUMBER_OF_ADAPTER_LEAF_IMPLEMENTATIONS, hardcodedClassNamesForAllKnownImplementations.size());

        for (String actualClassNameForAnImplementation : actualClassNamesForAllKnownImplementations) {
            assertThat(hardcodedClassNamesForAllKnownImplementations, hasItem(actualClassNameForAnImplementation));
        }
    }
}

File: Programmerare.CrsTransformations.Test/CrsTransformations/CrsTransformationAdapterLeafFactoryTest.cs
using Programmerare.CrsTransformations;

using System;
using System.Collections.Generic;
using NUnit.Framework;
using Programmerare.CrsConstants.ConstantsByAreaNameNumber.v9_5_4;
using Programmerare.CrsTransformations.Coordinate;
using Programmerare.CrsTransformations.Adapter.MightyLittleGeodesy;
using Programmerare.CrsTransformations.Adapter.DotSpatial;
using Programmerare.CrsTransformations.Adapter.ProjNet4GeoAPI;

[TestFixture]
public class CrsTransformationAdapterLeafFactoryTest {

    private CrsTransformationAdapterLeafFactory crsTransformationAdapterLeafFactory;

    private const int EXPECTED_NUMBER_OF_ADAPTER_LEAF_IMPLEMENTATIONS = CrsTransformationAdapterTest.EXPECTED_NUMBER_OF_ADAPTER_LEAF_IMPLEMENTATIONS;

    private static IList<string> actualClassNamesForAllKnownImplementations;
    
    [SetUp]
    public void SetUp() {
        crsTransformationAdapterLeafFactory = new CrsTransformationAdapterLeafFactory();
        actualClassNamesForAllKnownImplementations = new List<string> {
			typeof(CrsTransformationAdapterDotSpatial).FullName,
			typeof(CrsTransformationAdapterProjNet4GeoAPI).FullName,
			typeof(CrsTransformationAdapterMightyLittleGeodesy).FullName
        };
    }

    [Test]
    public void createCrsTransformationAdapter_shouldThrowException_whenTheParameterIsNotNameOfClassImplementingTheExpectedInterface() {
        string incorrectClassName = "abc";
        ArgumentException exception = Assert.Throws<ArgumentException>(() => {
            crsTransformationAdapterLeafFactory.CreateCrsTransformationAdapter(incorrectClassName);
        });

        string nameOfInterfaceThatShouldBeImplemented = typeof(ICrsTransformationAdapter).FullName;
        Assert.IsNotNull(exception);
        string exceptionMessage = exception.Message;
        Assert.That(exceptionMessage, Does.Contain(nameOfInterfaceThatShouldBeImplemented));
        Assert.That(exceptionMessage, Does.StartWith("Failed to instantiate")); // Fragile but the message string will not change often and if it does change then it will be very easy to modify the string here
    }

    [Test]
    public void listOfNonClassNamesForAdapters_shouldNotBeRecognizedAsAdapters()
    {
        List<string> stringsNotBeingClassNameForAnyAdapter = new List<string>() {
            null,
            "",
            "  ",
            " x ",
            "abc",
            // this test class i.e. the below "this" does not imlpement the interface so therefore assertFalse below
            this.GetType().FullName
        };

        foreach (string stringNotBeingClassNameForAnyAdapter in stringsNotBeingClassNameForAnyAdapter)
        {
            Assert.IsFalse(
                crsTransformationAdapterLeafFactory.IsCrsTransformationAdapter(stringNotBeingClassNameForAnyAdapter),
                "Should not have been recognized as adapter : " + stringNotBeingClassNameForAnyAdapter
            );
        }
    }

    [Test]
    public void listOfHardcodedClassnames_shouldBeCrsTransformationAdapters()
    {
        IList<string> hardcodedClassNamesForAllKnownImplementations = crsTransformationAdapterLeafFactory.GetClassNamesForAllKnownImplementations();
        foreach (string hardcodedClassNameForKnownImplementation in hardcodedClassNamesForAllKnownImplementations)
        {
            Assert.IsTrue(
                crsTransformationAdapterLeafFactory.IsCrsTransformationAdapter(hardcodedClassNameForKnownImplementation),
                "Name of failing class: " + hardcodedClassNameForKnownImplementation
            );
        }
    }

    [Test]
    public void listOfHardcodedClassnames_shouldBeCreateableAsNonNullCrsTransformationAdapters()
    {
        IList<String> hardcodedClassNamesForAllKnownImplementations = crsTransformationAdapterLeafFactory.GetClassNamesForAllKnownImplementations();
        foreach (string hardcodedClassNameForKnownImplementation in hardcodedClassNamesForAllKnownImplementations)
        {
            ICrsTransformationAdapter crsTransformationAdapter = crsTransformationAdapterLeafFactory.CreateCrsTransformationAdapter(hardcodedClassNameForKnownImplementation);
			VerifyThatTheCreatedAdapterIsRealObject(crsTransformationAdapter);
			Assert.That(actualClassNamesForAllKnownImplementations, Contains.Item(crsTransformationAdapter.LongNameOfImplementation));
        }
    }

    private void VerifyThatTheCreatedAdapterIsRealObject(ICrsTransformationAdapter crsTransformationAdapter)
    {
        Assert.IsNotNull(crsTransformationAdapter);
        // below trying to use the created object to really make sure it works
        CrsCoordinate coordinateWgs84 = CrsCoordinateFactory.CreateFromYNorthingLatitudeAndXEastingLongitude(59.330231, 18.059196, EpsgNumber.WORLD__WGS_84__4326);
        CrsTransformationResult resultSweref99 = crsTransformationAdapter.Transform(coordinateWgs84, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        Assert.IsNotNull(resultSweref99);
        Assert.IsTrue(resultSweref99.IsSuccess);
    }

    [Test]
    public void listOfKnownInstances_shouldOnlyContainNonNullObjectsAndTheNumberOfItemsShouldBeAtLeastFive()
    {
        IList<ICrsTransformationAdapter> list = crsTransformationAdapterLeafFactory.GetInstancesOfAllKnownAvailableImplementations();
        Assert.That(list.Count, Is.GreaterThanOrEqualTo(EXPECTED_NUMBER_OF_ADAPTER_LEAF_IMPLEMENTATIONS));
        foreach (ICrsTransformationAdapter crsTransformationAdapter in list)
        {
            VerifyThatTheCreatedAdapterIsRealObject(crsTransformationAdapter);
        }
    }

    [Test]
    public void listOfHardcodedClassnames_shouldCorrespondToActualClassNames()
    {
        IList<string> hardcodedClassNamesForAllKnownImplementations = crsTransformationAdapterLeafFactory.GetClassNamesForAllKnownImplementations();
        Assert.AreEqual(EXPECTED_NUMBER_OF_ADAPTER_LEAF_IMPLEMENTATIONS, hardcodedClassNamesForAllKnownImplementations.Count);
        Assert.AreEqual(EXPECTED_NUMBER_OF_ADAPTER_LEAF_IMPLEMENTATIONS, actualClassNamesForAllKnownImplementations.Count);

        foreach (string actualClassNameForAnImplementation in actualClassNamesForAllKnownImplementations)
        {
            Assert.That(hardcodedClassNamesForAllKnownImplementations, Contains.Item(actualClassNameForAnImplementation));
        }
        foreach (string hardcodedClassNamesForAllKnownImplementation in hardcodedClassNamesForAllKnownImplementations)
        {
            Assert.That(actualClassNamesForAllKnownImplementations, Contains.Item(hardcodedClassNamesForAllKnownImplementation));
        }
    }
}